source/module_hamilt_pw/hamilt_pwdft/parallel_grid.h

```diff
@@ -41,13 +51,13 @@ class Parallel_Grid
 	int ncz;
 	int ncxy;
 	int ncxyz;
-	int nczp; // different processors have different values.
+    int nczp; // number of z-layers (xy-planes) in each processor
 	int nrxx;
 	int nbz;
 	int bz;
 
-	bool allocate;
-    bool allocate_final_scf; //LiuXh add 20180619
+    bool allocate = false;
+    bool allocate_final_scf = false; //LiuXh add 20180619
```

Comment:
We should hear more from other developers

Reply:
> It seems that the number of write_cube input parameters are reduced. However, a new dependency is added to write_cube, which is the parallel_grid. I am not sure if the modification really makes the code a better code.

I think the cause of the dependency on Parallel_Grid is that
1. data passed to write_cube is in a distributed format in the presence of MPI;
2. data has to be collected to one process before writing the cube file;
3. the "collection" function is provided by Parallel_Grid (Parallel_Grid::reduce()).

If either of them can be avoided, the dependency can be removed. (Similar issues also exist for read_cube, where Parallel_Grid is already a dependency before this PR. The reason why Parallel_Grid was not a dependency of write_cube before is that the content of Parallel_Grid::reduce() was explicitly copied to write_cube_core to do the collection)

For item 1, one possible option is to collect data before calling write_cube, and pass the collected data to write_cube. I actually prefer this option, as it can make write_cube in module_io function much simpler. But the downside is that it increases the work where this function is called. Given that the reverse procedure "zpiece_to_all" in read_cube is more tedious than reduce, if similar change applies to read_cube, the amount of extra code would not be small.

Getting rid of item 2 is much more difficult, which requires MPI I/O. Since the amount of memory consumed by data is usually not an issue, I think this option can be safely ignored.

I think eliminating the duplicated content of Parallel_Grid::reduce() within write_cube_core is a good idea, but this does not necessarily introduce dependency on the class "Parallel_Grid". I think a possible option is to move "reduce" (and possibly zpiece_to_all" outside Parallel_Grid as a free function or member function of other class. I'm not a heavy user of Parallel_Grid and I'm not fully aware of the motivation of this design (it seems to carry the infomation to describe charge & some functions to help manipulate it - but why charge itself does not carry this information and capable of manipulating itself?). Therefore I'm not able to make any specific recommendation on this direction.